### PR TITLE
Attribute lock time to analyzers

### PR DESCRIPTION
**Pull request by sharwell**
Prior to this change, lock contention resulting from non-concurrent analyzers was not attributed to any specific analyzer. This change add the lock acquisition time to the analyzer execution time to help users identify analyzers responsible for slow builds, regardless of cause.

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs [hunk 1]
```diff
@@ -1564,6 +1582,13 @@ internal void ExecuteAndCatchIfThrows<TArg>(DiagnosticAnalyzer analyzer, Action<
             {
                 ExecuteAndCatchIfThrows_NoLock(analyzer, analyze, argument, info);
             }
+
+            if (_analyzerExecutionTimeMap != null)
+            {
+                var elapsed = timer.Elapsed.Ticks;
+                StrongBox<long> totalTicks = _analyzerExecutionTimeMap.GetOrAdd(analyzer, _ => new StrongBox<long>(0));
+                Interlocked.Add(ref totalTicks.Value, elapsed);
+            }
```
Comment (sharwell) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:1591]: 📝 Added code was just moved from the previous location.

Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:1591]: We should likely make `ExecuteAndCatchIfThrows_NoLock` a local function, so if in future there are more callsites added for it, the timing data still accounts for them.

**Review by mavasani — APPROVED** [review]

**Comment by 333fred** [comment]
Should we consider having multiple levels of measurement here? IE: this is the time it took your analyzer to execute. This additional time is the time we spent in lock contention because your analyzer isn't concurrent? Might help authors when troubleshooting why their analyzers are slow?

**Comment by sharwell** [comment]
@333fred piping that through would be difficult, and there is no scenario where an analyzer is expected to be non-concurrent (we have analyzers to remind analyzer authors, but some analyzers did not follow it).

**Review by 333fred — APPROVED** [review]

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:1591]: new and original code seem odd. should we have this in a finally?  seems weird to start a timer and not complete/report it in some cases (like cancellation)

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:1591]: (note: this is a non-blocking concern of mine as the original code had this issue too)

Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:1591]: > seems weird to start a timer and not complete/report it in some cases (like cancellation)

Results are not observable in cancellation cases.

**Review by jaredpar — APPROVED** [review]

**Pull request merged**